File: backend/src/main/java/com/hms/patient/service/PatientService.java
package com.hms.patient.service;

import com.hms.patient.dto.request.PatientRequestDTO;
import com.hms.patient.dto.response.PatientResponseDTO;
import org.springframework.data.domain.Slice;

import java.util.List;


public interface PatientService {

    PatientResponseDTO create(PatientRequestDTO dto);

    PatientResponseDTO getById(Long id);

    PatientResponseDTO update(Long id, PatientRequestDTO dto);

    void delete(Long id);

    List<PatientResponseDTO> getAll();

    Slice<PatientResponseDTO> getSlice(int page, int size);

    Slice<PatientResponseDTO> getSearchableSlice(int page, int size, String query);
}

File: frontend/src/app/features/patients/services/patient.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { ApiResponse, SliceResponse } from '../../../core/models/common.models';
import { Patient, PatientRequest } from '../models/patient.models';

@Injectable({
  providedIn: 'root',
})
export class PatientService {
  private apiUrl = `${environment.apiUrl}/patients`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<ApiResponse<Patient[]>> {
    return this.http.get<ApiResponse<Patient[]>>(this.apiUrl);
  }

  getSlice(page = 0, size = 20, query = ''): Observable<ApiResponse<SliceResponse<Patient>>> {
    const params: any = { page, size };
    if (query.trim()) params.query = query.trim();
    return this.http.get<ApiResponse<SliceResponse<Patient>>>(`${this.apiUrl}/slice`, { params });
  }

  getById(id: number): Observable<ApiResponse<Patient>> {
    return this.http.get<ApiResponse<Patient>>(`${this.apiUrl}/${id}`);
  }

  create(patient: PatientRequest): Observable<ApiResponse<Patient>> {
    return this.http.post<ApiResponse<Patient>>(this.apiUrl, patient);
  }

  update(id: number, patient: PatientRequest): Observable<ApiResponse<Patient>> {
    return this.http.put<ApiResponse<Patient>>(`${this.apiUrl}/${id}`, patient);
  }

  delete(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }
}
